packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -88,11 +88,160 @@ class AnnotationResult<T> {
+///
+///   @override
+///   bool get isRepaintBoundary => true; // The [layer] property will be used.
+///
+///   @override
+///   void paint(PaintingContext context, Offset offset) {
+///     _clipRectLayer.layer = context.pushClipRect(
+///       needsCompositing,
+///       offset,
+///       Offset.zero & size,
+///       super.paint,
+///       clipBehavior: Clip.hardEdge,
+///       oldLayer: _clipRectLayer.layer,
+///     );
+///   }
+///
+///   @override
+///   void dispose() {
+///     _clipRectLayer.layer = null;
+///     super.dispose();
+///   }
+/// }
+/// ```
+/// {@end-tool}
 /// See also:
 ///
 ///  * [RenderView.compositeFrame], which implements this recomposition protocol
 ///    for painting [RenderObject] trees on the display.
 abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// If asserts are enabled, returns whether [dispose] has
+  /// been called since the last time any retained resources were created.
+  ///
+  /// Throws an exception if asserts are disabled.
+  bool get debugDisposed {
+    late bool disposed;
+    assert(() {
+      disposed = _debugDisposed;
+      return true;
+    }());
+    return disposed;
+  }
+  // TODO(dnfield): https://github.com/flutter/flutter/issues/85066
+  final bool _debugDisposed = false;
+
+  /// Set when this layer is appended to a [ContainerLayer], and
+  /// unset when it is removed.
+  ///
+  /// This cannot be set from [attach] or [detach] which is called when an
+  /// entire subtree is attached to or detached from an owner. Layers may be
+  /// appended to or removed from a [ContainerLayer] regardless of whether they
+  /// are attached or detached, and detaching a layer from an owner does not
+  /// imply that it has been removed from its parent.
+  LayerHandle? _parentHandle;
+
+  /// Creates a handle that keeps the native resources of this layer from being
+  /// prematurely disposed.
+  ///
+  /// [RenderObject]s automatically store handles to [PictureLayer](s) they draw
+  /// into and dispose of those handles when the [RenderObject.dispose] method
+  /// is called or they otherwise no longer need the reference to that
+  /// [PictureLayer] (such as when they paint a new picture).
+  ///
+  /// A handle is also automatically managed for [RenderObject.layer].
+  ///
+  /// If a [RenderObject] creates layers in addition to its [RenderObject.layer],
+  /// it must create a handle to that layer and dispose of it when the layer is
+  /// no longer needed. For example, if it re-creates or nulls out an existing
+  /// layer in [RenderObject.paint], it should dispose of the handle to the
+  /// old layer. It should also dispose of any layer handles it holds in
+  /// [RenderObject.dispose].
+  LayerHandle createHandle() {
+    assert(!_debugDisposed, 'Attempted to create a handle to a disposed layer.');
+
+    _refCount += 1;
+    return LayerHandle._(this);
+  }
+
+  int _refCount = 0;
+
+  void _unref() {
+    assert(_refCount > 0);
```

Comment:
This would have to be a bug in the framework, if someone broke either `createHandle` or `LayerHandle.dispose`. End users should never see this, and if they somehow do they don't have a way to fix it short of patching the framework/filing a bug. Without a message here, they'll get a message encouraging them to file a bug with the stack trace/reproduction.